---
title: 'Ali''s RNAseq: All samples together: 01_createDDS'
output:
  html_document:
    df_print: paged
---
Script for creating dds objects of the subsetted data

#libraries
```{r message=FALSE, warning=FALSE}
library(DESeq2)
library(ggpubr)
library(rafalib)
library(pheatmap)
```

#folders and directories
```{r}
base.dir<- "/home/epicwl/c010-datasets/Internal/2018-Ali/RNASeq/190308_odcf/all_data"
data.dir <- file.path(base.dir, "data")
base_results.dir <- file.path(base.dir, "results")
results.dir<- file.path(base_results.dir , "tables")
PreDE.dir <- file.path(base_results.dir,"PreDE")
PostDE.dir <- file.path(base_results.dir,"PostDE")
featureCounts.dir <-"/home/epicwl/c010-datasets/Internal/2018-Ali/RNASeq/190308_odcf/all_data/results/repeats/"

```
#Data Read-in; create group annotation
```{r}
anno <- readRDS(file = file.path("/home/epicwl/c010-datasets/Internal/2018-Ali/RNASeq/190308_odcf/all_data/data","anno.rds"))
rownames(anno)<- anno$sample_name
anno$group <- as.factor(paste0(anno$genotype,"_Tam.",anno$tamoxifen_treatment ,"_Irr.", anno$irradiation_treatment, "_Hdac.",anno$hdac_treatment))
anno$replicate <- as.factor(anno$replicate)
```
#repeat procuedure for genes
#read in gene counts
```{r}
counts<- readRDS(file.path("/home/epicwl/c010-datasets/Internal/2018-Ali/RNASeq/190308_odcf/all_data/data", "counts.rds"))
```
#subset counts
```{r}
counts_genes <- counts
```
#create dds sets
```{r}
dds_genes <- DESeqDataSetFromMatrix(countData = counts_genes, 
                              colData = anno, 
                              design = ~ replicate + group)
#Estimating size factors
dds_genes <- estimateSizeFactors(dds_genes)

#Filter genes which are only expressed in 1 sample
length(dds_genes)
idx_dds_genes <- rowSums( counts(dds_genes, normalized=TRUE) >= 1 ) >= 1
table(idx_dds_genes)
dds_genes <- dds_genes
dim(dds_genes)

#Running the differential expression 
dds_genes <- DESeq(dds_genes)
saveRDS(dds_genes, file.path(results.dir, "dds_group_genes.rds"))
```


#raw data for counts and repeats

#raw data for counts and repeats
repeats=readRDS(c("/bigdisk/Nanopore/raw_data/repeats/repeats.RDS"))
count_repeats<- readRDS(file.path(featureCounts.dir, "merged_repeat_counts.rds"))

#make them integers
count_repeats<- sapply(count_repeats,function(x) as.integer(x))

#combine 
count_repeats<- cbind(as.data.frame(repeats),count_repeats)
#sum up repeat class family or
counts_sum_repFamily <- aggregate(cbind(count_repeats[,19:ncol(count_repeats)]), by=list(repFamily=count_repeats$repFamily), FUN=sum)
counts_sum_repClass <- aggregate(cbind(count_repeats[,19:ncol(count_repeats)]), by=list(repClass=count_repeats$repClass), FUN=sum)
counts_sum_repName <- aggregate(cbind(count_repeats[,19:ncol(count_repeats)]), by=list(repName=count_repeats$repName), FUN=sum)
repeat_list <- list(counts_sum_repFamily,counts_sum_repClass, counts_sum_repName )
names(repeat_list)<-  c("repFamily", "repClass", "repName")
saveRDS(repeat_list,file.path(results.dir, "repeat_list.rds"))

#create df of subsettet repeat namnes
count_repeats <- counts_sum_repName
rownames(count_repeats)<- counts_sum_repName$repName

#add sequencer to annotation
anno$sequencer <- "Illumina2000"
anno$sequencer <- ifelse(anno$replicate== 0, "Illumina2000", "NovaSeq")

#create dds
dds_repeats <- DESeqDataSetFromMatrix(countData = count_repeats[ anno$sample_name], 
                              colData = anno, 
                              design = ~ replicate + group)

#Estimating size factors
dds_repeats <- estimateSizeFactors(dds_repeats)

#look how many genes are expressed in at least one sample
length(dds_repeats)
idx_dds_repeats <- rowSums( counts(dds_repeats, normalized=TRUE) >= 1 ) >= 1
#table(idx_dds_repeats)
#dds <- dds
#dim(dds)

#Running the differential expression 
dds_repeats <- DESeq(dds_repeats)
saveRDS(dds_repeats, file.path(results.dir, "dds_group_repeats.rds"))

